src/core/server/elasticsearch/client/mocks.ts

```diff
@@ -31,42 +31,56 @@ const createInternalClientMock = (): DeeplyMockedKeys<Client> => {
     '_events',
     '_eventsCount',
     '_maxListeners',
+    'constructor',
     'name',
     'serializer',
     'connectionPool',
     'transport',
     'helpers',
   ];
 
+  const getAllPropertyDescriptors = (obj: Record<string, any>) => {
+    const descriptors = Object.entries(Object.getOwnPropertyDescriptors(obj));
+    let prototype = Object.getPrototypeOf(obj);
+    while (prototype != null && prototype !== Object.prototype) {
+      descriptors.push(...Object.entries(Object.getOwnPropertyDescriptors(prototype)));
+      prototype = Object.getPrototypeOf(prototype);
+    }
+    return descriptors;
+  };
+
   const mockify = (obj: Record<string, any>, omitted: string[] = []) => {
-    Object.keys(obj)
-      .filter((key) => !omitted.includes(key))
-      .forEach((key) => {
-        const propType = typeof obj[key];
-        if (propType === 'function') {
+    const descriptors = getAllPropertyDescriptors(obj);
+    descriptors
+      .filter((descriptor) => !omitted.includes(descriptor[0]))
+      .forEach((descriptor) => {
+        const key = descriptor[0];
+        if (typeof descriptor[1].value === 'function') {
           obj[key] = jest.fn(() => createSuccessTransportRequestPromise({}));
-        } else if (propType === 'object' && obj[key] != null) {
-          mockify(obj[key]);
+        } else if (typeof obj[key] === 'object' && obj[key] != null) {
+          mockify(obj[key], omitted);
         }
       });
   };
 
   mockify(client, omittedProps);
 
-  // client got some read-only (getter) properties
-  // so we need to extend it to override the getter-only props.
-  const mock: any = { ...client };
-
-  mock.transport = {
+  client.close = jest.fn().mockReturnValue(Promise.resolve());
+  client.child = jest.fn().mockImplementation(() => createInternalClientMock());
+  Object.defineProperty(client, 'on', {
```

Comment:
Why did you use the different forms for `on`, `off`, and `once` declaration? To not allow to re-define them, maybe?

Reply:
`on`, `off`, and `once` are properties without a setter, so if you try to do `client.on = jest.fn()` it throws the following error: `TypeError: Cannot set property on of #<Client> which has only a getter`. I can add a comment to clarify why we're doing this.

Reply:
so a developer cannot define a custom mock implementation, can one? ok, let's keep it as is...YAGNI